src/charm.py

```diff
@@ -24,22 +28,80 @@ def __init__(self, *args: typing.Any) -> None:
             args: passthrough to CharmBase.
         """
         super().__init__(*args)
+        self._charm_state = CharmState(charm=self)
+        self._webserver = GunicornWebserver(charm_state=self._charm_state)
         self.framework.observe(self.on.config_changed, self.config_service)
 
+    def flask_container(self, require_connected: bool = True) -> Container:
+        """Get the flask application workload container controller.
+
+        Args:
+            require_connected: if set to ``True``, a runtime exception will be raised if the
+                pebble inside the container is not ready.
+
+        Return:
+            The controller of the flask application workload container.
+
+        Raises:
+            RuntimeError: if the pebble inside  the container is not ready while the
+                ``require_connected`` is set to True.
+        """
+        container = self.unit.get_container("flask-app")
+        if require_connected and not container.can_connect():
+            raise RuntimeError("pebble inside flask-app container is not ready")
+        return container
+
     def config_service(self, event: ConfigChangedEvent) -> None:
         """Configure the flask pebble service layer.
 
         Args:
             event: the config-changed event that trigger this callback function.
         """
-        container = self.unit.get_container("flask-app")
+        container = self.flask_container(require_connected=False)
         if not container.can_connect():
             event.defer()
             return
+
+        service_name = "flask-app"
         container.add_layer("flask-app", self.flask_layer(), combine=True)
+        webserver_config_path = str(self._webserver.config_path)
+        current_webserver_config = self.pull_file(webserver_config_path)
+        is_webserver_running = container.get_service(service_name).is_running()
+        if current_webserver_config != self._webserver.config:
```

Comment:
If the configuration has changed, it need to send a signal to the Gunicorn server to reload the configuration. The reload process is a little disruptive to the Gunicorn server, so I would like to eliminate the unnecessary reloads.

Reply:
Can't we let the webserver service define if a reload is needed ?
This will let the service define how to determine if event triggered requires a signal to be sent and let the charm send the signal using pebble client

Reply:
If the webserver service can can determine if a reload is needed, the webserver service is not purely about the webserver anymore, it will be a web server pebble service. It will break the single-responsibility principle.

Reply:
I don't agree, defining if a reload is needed is purely the webserver responsibility, some other webserver could require a restart in other condition therefore we need an interface to specify if a reload is needed.
I don't expect the webserver service to effectively do the reload, but to determine if a reload is needed or not depending on the state changes.
Therefore having a "should_reload" or "must_reload" method in the webserver service that will implement the logic and return true if a reload is needed is the way to go.
And then the charm execute the reload logic through Pebble

Reply:
Indeed some web server require a restart in other condition, but if the `FlaskCharm` doesn't know when this "other condition" would happen, the `FlaskCharm` wouldn't have the chance to call the `should_reload` or `must_reload`. We can't just call the `should_reload` and `must_reload` between every line. This is why the `FlaskCharm` must know when the web server should reload in order to reload the web server.

Reply:
The problem is that the charm has to know when the webserver needs to reload to know which place should the snippet `if self._webserver.should_reload(current_webserver_config)...` be put inside the charm. And if the charm already know when the webserver needs to reload, the charm can just reload by it self.

Reply:
Let's make a call, I think we don't understand each other :D
What you say is not incompatible with what I say:
It's the charm that does the reload in my example, it's the charm that decides when the reload is done in my example, but it's the webserver service that knows if a reload is needed

Reply:
I updated the code to use webserver service to decide the reload based on your comment, thanks!